storage/minifile/mimifile.go

```diff
@@ -0,0 +1,307 @@
+package minifile
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/prometheus/tsdb/fileutil"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type MiniFile struct {
+	path         string
+	instanceLock fileutil.Releaser // File-system lock to prevent double opens
+	namedLock    *NamedLock
+	lock         *sync.RWMutex
+	closed       int64
+}
+
+type NamedLock struct {
+	lock  *sync.RWMutex
+	locks map[string]*sync.RWMutex
+}
+
+func NewNamedLock() *NamedLock {
+	return &NamedLock{
+		lock:  &sync.RWMutex{},
+		locks: make(map[string]*sync.RWMutex),
+	}
+}
+
+func New(path string) (*MiniFile, error) {
+	abs, err := filepath.Abs(path)
+	if err != nil {
+		return nil, err
+	}
+
+	flock, _, err := fileutil.Flock(filepath.Join(abs, "FLOCK"))
+	if err != nil {
+		return nil, err
+	}
+
+	err = os.MkdirAll(abs, 0755)
+	if err != nil {
+		return nil, err
+	}
+	return &MiniFile{
+		path:         abs,
+		instanceLock: flock,
+		lock:         &sync.RWMutex{},
+		namedLock:    NewNamedLock(),
+	}, nil
+}
+
+func (mf *MiniFile) Put(key string, value []byte) error {
+	if mf.isClosed() {
+		return fmt.Errorf("the miniFile storage is closed")
+	}
+
+	if key == "" {
+		return fmt.Errorf("store file with empty key")
+	}
+
+	crc := make([]byte, 4)
+	binary.LittleEndian.PutUint32(crc, util.NewCRC(value).Value())
+	value = append(value, crc...)
+
+	mf.lock.RLock()
+	defer mf.lock.RUnlock()
+
+	mf.namedLock.Lock(key)
+	defer mf.namedLock.UnLock(key, false)
+
+	name := filepath.Join(mf.path, key)
+
+	if err := ioutil.WriteFile(name, value, 0644); err != nil {
+		return fmt.Errorf("fail to write file %s: %w", name, err)
+	}
+
+	return nil
+}
+
+func (mf *MiniFile) Delete(key string) error {
+	if mf.isClosed() {
+		return fmt.Errorf("the miniFile storage is closed")
+	}
+
+	mf.lock.RLock()
+	defer mf.lock.RUnlock()
+
+	mf.namedLock.Lock(key)
+	defer mf.namedLock.UnLock(key, true)
+
+	err := os.Remove(filepath.Join(mf.path, key))
+	if err != nil && isNoFileError(err) {
+		return nil
+	}
+
+	return err
+}
+func (mf *MiniFile) Get(key string) ([]byte, error) {
+	if mf.isClosed() {
+		return nil, fmt.Errorf("the miniFile storage is closed")
+	}
+
+	mf.lock.RLock()
+	defer mf.lock.RUnlock()
+
+	return mf.get(key)
+}
+
+func (mf *MiniFile) get(key string) ([]byte, error) {
+	mf.namedLock.RLock(key)
+
+	name := filepath.Join(mf.path, key)
+	val, err := ioutil.ReadFile(name)
+	if err != nil && isNoFileError(err) {
+		_ = mf.namedLock.RUnLock(key, true)
+		return nil, nil
+	}
+
+	_ = mf.namedLock.RUnLock(key, false)
+
+	crc := make([]byte, 4)
+	binary.LittleEndian.PutUint32(crc, util.NewCRC(val[:len(val)-4]).Value())
```

Comment:
Please check that the length of val is greater than 4

Reply:
fixed